src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,350 @@
+        wavelet_name: str = 'db4',
+        level: int | None = None,
+    ):
+        """Wavelet operator.
+
+        Parameters
+        ----------
+        domain_shape
+            Shape of domain where wavelets are calculated. If set to None the shape is taken from the input of the
+            forward operator. The adjoint operator will raise an error.
+        dim
+            Dimensions (axes) where wavelets are calculated
+        wavelet_name
+            Name of wavelets
+        level
+            Highest wavelet level. If set to None, the highest possible level is calculated based on the domain shape.
+
+        Raises
+        ------
+        ValueError
+            If wavelets are calculated for more than three dimensions.
+        ValueError
+            If wavelet dimensions and domain shape do not match.
+        NotImplementedError
+            If the domain shape is odd. Adjoint will lead to the wrong domain shape.
+        """
+        super().__init__()
+        self.domain_shape = domain_shape
+        self.wavelet_name = wavelet_name
+        self.dim = dim
+        self.level = level
+
+        # number of wavelet directions
+        if len(dim) == 1:
+            self.n_wavelet_directions = 1
+        elif len(dim) == 2:
+            self.n_wavelet_directions = 3
+        elif len(dim) == 3:
+            self.n_wavelet_directions = 7
+        else:
+            raise ValueError('Only 1D, 2D and 3D wavelet transforms are supported.')
+
+        if domain_shape is not None:
+            if len(dim) != len(domain_shape):
+                raise ValueError(
+                    'Number of dimensions along which the wavelet transform should be calculated needs to',
+                    'be the same as the domain shape',
+                )
+
+            if any(d % 2 for d in domain_shape):
+                raise NotImplementedError('ptwt only supports wavelet transforms for even number of samples.')
+
+            # size of wavelets
+            wavelet_length = torch.as_tensor((Wavelet(wavelet_name).dec_len,) * len(domain_shape))
+
+            # calculate shape of wavelet coefficients at each level
+            current_shape = torch.as_tensor(domain_shape)
+
+            if _check_level(domain_shape, wavelet_length, level) == 0:
+                self.coefficients_shape = [current_shape]
+                self.coefficients_padding = [(0, 0)] * len(current_shape)
+            else:
+                self.coefficients_shape = []
+                self.coefficients_padding = []
+                for _ in range(_check_level(domain_shape, wavelet_length, level)):
+                    # Add padding
+                    for ind in range(len(current_shape)):
+                        padl, padr = _get_pad(current_shape[ind], wavelet_length[ind])
+                        current_shape[ind] += padl + padr
+                    current_shape = torch.floor((current_shape - (wavelet_length - 1) - 1) / 2 + 1).to(
```

Comment:
suggestion:
(current_shape - wavelet_length) / 2 + 1).to(torch.int64)

Reply:
combining (see comment in _get_pad)
# pytorch wavelet toolbox pads such that the output is (data_len + filt_len - 1)//2, matching pywt
((current_shape + wavelet_length- 1)/2).int64()